src/states/game.gd

```diff
@@ -209,27 +214,43 @@ func spawn_peer_player(player_id: int) -> void:
 	player.get_node("Head/HeadMesh").set_material_override(material)
 	player.set_multiplayer_authority(player_id)
 	$Players.add_child(player)
+	if is_multiplayer_authority():
+		player.player_death.connect(assign_spawn_point.bind(player_id))
+		player.player_spawn.connect(clear_spawn_point.bind(player_id))
+
+
+# Assign a spawn point to a player, if one does not exist. Called only on the server.
+func assign_spawn_point(player_id: int) -> void:
+	if not is_multiplayer_authority():
+		return
+	print("Assigning spawn point to player %d" % player_id)
+	var spawn_point: SpawnPoint
+	# Check if there are any pre-assigned spawn points for this player.
+	var assigned_spawn_points := spawn_points.filter(func(x): return x.player_id == player_id)
+	if len(assigned_spawn_points) > 0:
+		spawn_point = assigned_spawn_points.pick_random()
+	else:
+		var available_spawn_points := spawn_points.filter(func(x): return x.available(player_id))
+		spawn_point = available_spawn_points.pick_random() as SpawnPoint
```

Comment:
Should probably add in a condition in the very rare case that there are also no available spawn points. In that case we just gotta accept that players are gonna spawn on top of each other
```suggestion
		var available_spawn_points := spawn_points.filter(func(x): return x.available(player_id))
		if len(available_spawn_points) == 0:
			available_spawn_points = spawn_points
		spawn_point = available_spawn_points.pick_random() as SpawnPoint
```

Reply:
Ooh, I've never seen this suggest change feature, seems cool. I agree